homeassistant/components/lifx/light.py

```diff
@@ -377,68 +384,70 @@ async def start_effect(self, entities, service, **kwargs):
             await self.effects_conductor.stop(bulbs)
 
     @callback
-    def register(self, bulb):
-        """Allow a single in-flight discovery per bulb."""
-        if bulb.mac_addr not in self.discoveries_inflight:
-            self.discoveries_inflight[bulb.mac_addr] = bulb.ip_addr
-            _LOGGER.debug("Discovered %s (%s)", bulb.ip_addr, bulb.mac_addr)
-            self.hass.async_create_task(self.register_bulb(bulb))
-        else:
-            _LOGGER.warning("Duplicate LIFX discovery response ignored")
+    def register(self, bulb: Light) -> None:
+        """Try and skip processing LIFX Switches as early as possible."""
+        if bulb.mac_addr not in self.switch_devices:
+            self.hass.async_create_task(self._handle_discovery(bulb))
```

Comment:
As long as there are multiple discoveries feeding in data there is sure looks like going to be a race here without a lock

Possible scenerio:

Discovery A sees the device and calls `register`, the task is created and control is returned to the event loop
Discovery B sees the device and calls `register`, the task is created and control is returned to the event loop
The `_handle_discovery` call for A is called and proceeds to `await ack(bulb.get_color)` and control is return to the event loop
The `_handle_discovery` call for B is called and proceeds to `await ack(bulb.get_color)` and control is return to the event loop
The `_handle_discovery` call for A now proceeds after the `ack` await finishes for A is called and proceeds to create the entity and add it
The `_handle_discovery` call for B now proceeds after the `ack` await finishes for A is called and proceeds to create the entity and add a duplicate entity.

Reply:
Yes, I know. The problem is, folks have come to rely on this behaviour because it allows the integration to discover devices significantly faster and more reliably. With any sort of in-flight check, the whole thing just takes way too long (or never completes).